### PR TITLE
feat: Add eth_getBlockTransactionCountByHash and eth_getBlockTransactionCountByNumber

### LINKED ISSUE #43: eth_api: Implement `get_block_transaction_count_by_number` method
### **Description:**
We need to implement the [`get_block_transaction_count_by_number`](https://github.com/matter-labs/era-test-node/blob/3539ad0a87f5cc5350b346bace874f4d966714b3/src/node.rs#L1028C8-L1028C45) method. This function returns the number of transactions in the block with the given block number.

```rust
/// Retrieves the number of transactions in a block by block number.
fn get_block_transaction_count_by_number(&self, block_number: BlockNumber) -> jsonrpc_core::BoxFuture<jsonrpc_core::Result<U256>> {
    not_implemented("get_block_transaction_count_by_number")
}
```

#### Expected response
Command:
```
# Note: Using mainnet URL for demonstration purposes
curl -X POST https://mainnet.era.zksync.io \
     -H "Content-Type: application/json" \
     --data '{
       "jsonrpc": "2.0",
       "id": 1,
       "method": "eth_getBlockTransactionCountByNumber",
       "params": ["0xA4CDA7"]
     }'
```
Response:
```
{"jsonrpc":"2.0","result":"0x2","id":1}
```

### **Requirements:**

1. **Implement the Method:**
    - The method should take in a `BlockNumber`.
    - It should return the count of transactions in the specified block as a `U256`.

2. **Unit Tests:**
    - Write unit tests ensuring the `get_block_transaction_count_by_number` method functions correctly.
    - Test various edge cases, including blocks with no transactions.
    - Properly handle potential exceptions and errors.

3. **Inline Documentation:**
    - Write inline Rust documentation (`///` comments) explaining the purpose and usage of the `get_block_transaction_count_by_number` method.
    - Include details on parameters, return values, and any possible exceptions or errors.

### **Task Checklist:**
- [ ] Implement the `get_block_transaction_count_by_number` method.
- [ ] Write unit tests for the `get_block_transaction_count_by_number` method.
- [ ] Add inline documentation for the `get_block_transaction_count_by_number` method.

### **Additional Notes:**
If you're an external contributor looking to take on this issue, please comment below to let us know you're working on it. If you need further clarification on the requirements or need assistance with anything related to the implementation, feel free to ask!

### LINKED ISSUE #44: eth_api: Implement `get_block_transaction_count_by_hash` method
### **Description:**
We are tasked with implementing the [`get_block_transaction_count_by_hash`](https://github.com/matter-labs/era-test-node/blob/3539ad0a87f5cc5350b346bace874f4d966714b3/src/node.rs#L1035C8-L1035C43) method. This function returns the number of transactions in the block with the given block hash.

```rust
/// Retrieves the number of transactions in a block by block hash.
fn get_block_transaction_count_by_hash(&self, block_hash: H256) -> jsonrpc_core::BoxFuture<jsonrpc_core::Result<U256>> {
    not_implemented("get_block_transaction_count_by_hash")
}
```

#### Expected response

Command:
```
# Note: Using mainnet URL for demonstration purposes
curl -X POST https://mainnet.era.zksync.io \
     -H "Content-Type: application/json" \
     --data '{
       "jsonrpc": "2.0",
       "id": 1,
       "method": "eth_getBlockTransactionCountByHash",
       "params": ["0x3e72b04fda2b20d09bc97380dfdbd8b9e78272829cd9b337a064ab533bc6502d"]
     }'
```
Response:
```
{"jsonrpc":"2.0","result":,"id":1}
```

### **Requirements:**

1. **Implement the Method:**
    - The method should accept a `H256` block hash as its parameter.
    - It should then return the count of transactions present in the block corresponding to the given hash, formatted as a `U256`.

2. **Unit Tests:**
    - Write unit tests ensuring the `get_block_transaction_count_by_hash` method behaves as expected.
    - Test a variety of edge cases, including blocks without transactions.
    - Include any relevant information about the return values and potential errors/exceptions.

3. **Inline Documentation:**
    - Add inline Rust documentation (`///` comments) to describe the functionality and usage of the `get_block_transaction_count_by_hash` method.
    - Include any relevant information about the return values and potential errors/exceptions.

### **Task Checklist:**
- [ ] Implement the `get_block_transaction_count_by_hash` method.
- [ ] Write unit tests for the `get_block_transaction_count_by_hash` method.
- [ ] Add inline documentation for the `get_block_transaction_count_by_hash` method.

### **Additional Notes:**
If you're an external contributor looking to take on this issue, please comment below to let us know you're working on it. If you need further clarification on the requirements or need assistance with anything related to the implementation, feel free to ask!

### PR DESCRIPTION
**Pull request by nbaztec**
# What :computer: 
* Adds support for `eth_getBlockTransactionCountByHash`
* Adds support for `eth_getBlockTransactionCountByNumber`
* Improves `testing` server

# Why :hand:
* Users might need to view the number of transactions contained in a block by its hash or by the number.
* It is beneficial to generate raw json for blocks and transactions instead of the response object during testing as many objects are composable.

# Evidence :camera:
image

# Notes :memo:
* Fixes #43 
* Fixes #44

### CODE REVIEW

File: src/node.rs [hunk 1]
```diff
@@ -1832,16 +1832,130 @@ impl<S: Send + Sync + 'static + ForkSource + std::fmt::Debug> EthNamespaceT for
...
-        not_implemented("get_block_transaction_count_by_number")
+        let inner = Arc::clone(&self.inner);
+
+        Box::pin(async move {
+            let maybe_result = {
+                let reader = match inner.read() {
+                    Ok(r) => r,
+                    Err(_) => return Err(into_jsrpc_error(Web3Error::InternalError)),
+                };
+                match block_number {
```
Comment (MexicanAce) [inline src/node.rs:1845]: suggestion: Feels like we can reduce the amount of code here by calculating the block_hash first:

```rust
let block_hash = match block_number {
    zksync_types::api::BlockNumber::Latest
    | zksync_types::api::BlockNumber::Pending
    | zksync_types::api::BlockNumber::Finalized
    | zksync_types::api::BlockNumber::Committed => reader
        .block_hashes
        .get(&reader.current_miniblock),
    zksync_types::api::BlockNumber::Number(ask_number) => reader
        .block_hashes
        .get(&ask_number.as_u64()),
    zksync_types::api::BlockNumber::Earliest => reader
        .block_hashes
        .get(&0),
};

reader
    .blocks
    .get(block_hash)
    .map(|block| U256::from(block.transactions.len()))
    .or_else(|| {
        reader
            .fork_storage
            .inner
            .read()
            .expect("failed reading fork storage")
            .fork
            .as_ref()
            .and_then(|fork| {
                fork.fork_source
                    .get_block_transaction_count_by_number(block_number)
                    .ok()
                    .flatten()
            })
    })
```

File: src/node.rs [hunk 2]
```diff
@@ -1970,35 +2084,8 @@ mod tests {
     #[tokio::test]
     async fn test_get_block_by_hash_for_produced_block() {
         let node = InMemoryNode::<HttpForkSource>::default();
+        let expected_block_hash = testing::apply_tx(&node);
```
Comment (MexicanAce) [inline src/node.rs:2087]: Thanks for this!! Was going to suggest something similar in a previous PR 😅

**Review by MexicanAce — APPROVED** [review]
LGTM, left 1 suggestion

**Review by MexicanAce — APPROVED** [review]
LGTM
